### PR TITLE
Mwilson/reactor stats WIP

### PR DESCRIPTION
**Pull request by geekinutah**
### What does this PR do?
Adds stat tracking to ThreadPool used for Reactor. Periodically fires a reactor_threadpool_stats event.

### What issues does this PR fix or reference?
46431

### Previous Behavior
No stat tracking

### New Behavior
See What does this PR do

### Tests written?

Not yet

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by geekinutah** [comment]
@cachedout Pls read over and let me know if this is what you had in mind.

**Comment by rallytime** [comment]
@geekinutah There's some lint issues that need to be addressed with this change.

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48874/1/warnings52Result/

File: salt/utils/process.py [hunk 1]
```diff
@@ -299,13 +309,35 @@ def __init__(self,
 
         self._workers = []
 
+        # keep stats on overflows, puts, tasks_done
+        self._stats = {'overflows': 0,
+                       'puts': 0,
+                       'tasks_done': 0
+                      }
+
+        self.event = salt.utils.event.get_event(
```
Comment (cachedout) [inline salt/utils/process.py]: Definitely only want to do this if a configuration option is set to gather stats.

**Review by cachedout — COMMENTED** [review]
This is exactly the sort of approach I was envisioning. Well played!

**Comment by cachedout** [comment]
@geekinutah Just wondering if you saw the above feedback and if there are any other questions that I can answer.

**Review by cachedout — APPROVED** [review]

**Comment by geekinutah** [comment]
@cachedout I saw your feedback, thanks. I will re-write this addressing lint and some other obvious issues. What are your thoughts on tying this stats reporting to the enable_master_stats and master_stats_interval configs?

**Comment by cachedout** [comment]
@geekinutah I think that's a good idea but to accomplish that I think that we need to break apart `master_stats` so that a user can switch on and off different types of stats at-will. If we can achieve that, which shouldn't be hard, I think we'd be in awesome shape!

File: salt/utils/process.py [hunk 2]
```diff
@@ -288,7 +290,15 @@ class ThreadPool(object):
     '''
     def __init__(self,
                  num_threads=None,
-                 queue_size=0):
+                 queue_size=0,
+                 opts=None):
+        if not opts:
```
Comment (cachedout) [inline salt/utils/process.py]: I almost think we should make `opts` a required arg here. We should already have it generated by a caller anyway.

**Comment by cachedout** [comment]
@geekinutah Some small lint errors here:

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48874/3/warnings52Result/new/

File: salt/utils/process.py (hunk 2 above)
Comment (geekinutah) [reply salt/utils/process.py]: Okay, I think it sounds long term'ish to get a nice consistent paradigm for stats. For now I will keep this discrete for simplicity. I should be able to push an update to this soon :-).

**Comment by gtmanfred** [comment]
@geekinutah can you update these lint errors? it looks like that is what is preventing the tests from running.

File: salt/utils/process.py (hunk 2 above)
Comment (cachedout) [reply salt/utils/process.py]: OK, sounds good!

**Comment by rallytime** [comment]
@geekinutah It looks like you were planning on pushing an update to this, based on your most recent comment. Were you able to come back to this?

**Comment by geekinutah** [comment]
Sorry @rallytime, I haven't been able to get back to it yet. Feel free to finish, it might be a while (weeks) before I can finish up.

**Comment by rallytime** [comment]
@geekinutah No problem at all. I will close this for now, and then when you're ready to come back to it, we can open it up again. Just let us know when you're ready. :)

**Pull request closed without merging**